### PR TITLE
Link task manager, page manager and PGD/PTB

### PR DESCRIPTION
**Pull request by Zefinder**

### CODE REVIEW

File: kernel/core/extend/pagemem.c [hunk 1]
```diff
@@ -59,3 +61,77 @@ void init_pgd()
...
+            usr_forced_map(&usr_PGD[0], 0x4000000, SHARED_START, PTE_OFFSET);                   // map shared memory (1 page)
+			break;
+		default:
+			break;
+	}
+
+    krn_identity_map(&krn_PGD[0], SHARED_START, SHARED_END + PTE_OFFSET - 1); // map shared usr page for Kernel
+}
+
+void task_forced_map(tidx task, offset_t virtual_address, offset_t physical_address, offset_t size)
```
Comment (angeredShrimp) [inline kernel/core/extend/pagemem.c]: maybe specify virtual_address_start, that or specify in function description

File: kernel/core/extend/pagemem.c [hunk 2]
```diff
@@ -59,3 +61,77 @@ void init_pgd()
...
+{
+	pde32_t* krn_pgd = nth_pgd_gbl(0);
+	pde32_t* usr_pgd = nth_user_pgds(task);
+
+	// Put in user ptb
+    usr_forced_map(usr_pgd, virtual_address, physical_address, size);
+
+    // Put in kernel ptb
+    krn_identity_map(krn_pgd, physical_address, physical_address + size);
```
Comment (angeredShrimp) [inline kernel/core/extend/pagemem.c]:
```suggestion
    krn_identity_map(krn_pgd, physical_address, physical_address + size - 1);
```
end address is included in mapping

File: kernel/core/extend/task_manager.c [hunk 3]
```diff
@@ -44,6 +48,7 @@ int create_task(void) {
     
     // If address is NULL then repush the task index and we return 1
     if (first_page_index == 0ul) {
+        // TODO Remove PGD
```
Comment (angeredShrimp) [inline kernel/core/extend/task_manager.c]: clear_task_pagemem then ?

File: kernel/core/extend/userland.c [hunk 4]
```diff
@@ -3,7 +3,7 @@
 #include <extend/userland.h>
 #include <extend/intr.h>
 
-void __attribute__((section(".usr"),aligned(4))) userland()
+void __attribute__((section(".task2"),aligned(4))) userland()
```
Comment (angeredShrimp) [inline kernel/core/extend/userland.c:6]: Why change it ? Leave it in usr if you don't need it, unless for testing I guess.

File: kernel/include/extend/pagemem.h [hunk 5]
```diff
@@ -150,18 +150,49 @@
...
+// Fills task PTB with the virtual and physical addresses
+void task_forced_map(tidx task, offset_t virtual_address, offset_t physical_address, offset_t size);
+
+// Flushes paging entries in usr_ptb and mirrors actions in krn_ptb
+void flush_ptb(pte32_t *usr_ptb, pte32_t *krn_ptb);
+
+// Flushes paging entries in usr_pgd and mirrors actions in krn_pgd
+void flush_pgd(pde32_t *usr_pgd, pde32_t *krn_pgd);
+
+void clear_task_pagemem(tidx task);
```
Comment (angeredShrimp) [inline kernel/include/extend/pagemem.h:197]: // Flushes task's PGD and PTBs, clears corresponding entries in kernel's PGD

File: kernel/include/extend/userland.h [hunk 6]
```diff
@@ -1,6 +1,6 @@
 #ifndef __EXTEND_USERLAND_H__
 #define __EXTEND_USERLAND_H__
 
-void __attribute__((section(".usr"),aligned(4))) userland(void);
+void __attribute__((section(".task2"),aligned(4))) userland(void);
```
Comment (angeredShrimp) [inline kernel/include/extend/userland.h:4]: same

File: kernel/include/pagemem.h [hunk 7]
```diff
@@ -161,7 +161,7 @@ typedef union page_table_entry
 #define __copy_page(_d,_s)           _memcpy32(_d, _s, PAGE_SIZE)
 
 /*
-** Invalidate 32 bits TLB entry
+** Invalidate 32 bits TLB entry, only for TLB
```
Comment (angeredShrimp) [inline kernel/include/pagemem.h:164]: Haha right, almost fell for it

**Review by angeredShrimp — CHANGES_REQUESTED** [review]

File: kernel/core/extend/userland.c (hunk 4 above)
Comment (Zefinder) [reply kernel/core/extend/userland.c:6]: Bro is reacting to what he has done in his core_polish that I merged...

File: kernel/include/extend/userland.h (hunk 6 above)
Comment (Zefinder) [reply kernel/include/extend/userland.h:4]: Same

File: kernel/include/pagemem.h (hunk 7 above)
Comment (Zefinder) [reply kernel/include/pagemem.h:164]: Bro is still reacting to what he has done in his core_polish that I merged...

File: kernel/core/extend/task_manager.c (hunk 3 above)
Comment (Zefinder) [reply kernel/core/extend/task_manager.c]: Yes

**Review by angeredShrimp — APPROVED** [review]

**Pull request merged**